# Sesión 07 · Encuentra lo que necesitas

Consulta filas por etiquetas o posiciones y formula filtros que respondan preguntas.

Duración: 150 minutos de clase + 45–60 minutos de tarea. Ejecuta en orden. Los ejemplos son independientes salvo indicación.

## Objetivos
- Distinguir loc e iloc.
- Combinar condiciones correctamente.
- Crear columnas y ordenar resultados.

## loc e iloc: dos coordenadas distintas

loc accede por etiquetas y acepta condiciones booleanas. iloc accede por posiciones enteras desde cero. Una etiqueta puede ser texto o un número; no debe confundirse con una posición.

En el índice único y ordenado de este ejemplo, loc incluye la etiqueta final de un rango. iloc excluye la posición final.

In [1]:
import pandas as pd

productos = pd.DataFrame({
    "Producto": ["Mouse", "Teclado", "Monitor"],
    "Precio": [15.99, 25.50, 199.99],
    "Cantidad": [3, 2, 1]
}, index=["p1", "p2", "p3"])
print(productos.loc["p2", "Precio"])
print(productos.iloc[1, 1])
print(productos.loc["p1":"p2"])
print(productos.iloc[0:2])

25.5
25.5
   Producto  Precio  Cantidad
p1    Mouse   15.99         3
p2  Teclado   25.50         2
   Producto  Precio  Cantidad
p1    Mouse   15.99         3
p2  Teclado   25.50         2


**Qué observar:** Ambos accesos individuales devuelven 25.5; ambos rangos seleccionan Mouse y Teclado.

**Recuerda:** Tras ordenar una tabla, la posición de una fila puede cambiar mientras su etiqueta permanece.

## Filtros: convierte una pregunta en condición

Una comparación de columna produce una máscara booleana. Para combinar condiciones en Pandas usa & para “y”, | para “o”, y paréntesis alrededor de cada comparación.

Filtrar devuelve los registros que cumplen la regla. Especificar las columnas con loc permite mostrar solo la información necesaria.

In [2]:
import pandas as pd

productos = pd.DataFrame({
    "Producto": ["Mouse", "Teclado", "Monitor"],
    "Precio": [15.99, 25.50, 199.99],
    "Cantidad": [3, 2, 1]
}, index=["p1", "p2", "p3"])
condicion = (productos["Precio"] < 100) & (productos["Cantidad"] >= 2)
print(productos.loc[condicion, ["Producto", "Cantidad"]])

   Producto  Cantidad
p1    Mouse         3
p2  Teclado         2


**Qué observar:** Mouse con 3 unidades y Teclado con 2.

**Recuerda:** En condiciones de Pandas no sustituyas & por and: and no combina Series elemento por elemento.

## Derivar, modificar y ordenar

Una nueva columna puede calcularse a partir de columnas existentes. Para modificar valores según una condición, usa una sola operación loc y evita asignaciones encadenadas.

sort_values crea una tabla ordenada por valores. Guardamos el resultado en otra variable para conservar la referencia original.

In [3]:
import pandas as pd

productos = pd.DataFrame({
    "Producto": ["Mouse", "Teclado", "Monitor"],
    "Precio": [15.99, 25.50, 199.99],
    "Cantidad": [3, 2, 1]
}, index=["p1", "p2", "p3"])
productos["Ingreso"] = productos["Precio"] * productos["Cantidad"]
productos.loc[productos["Producto"] == "Mouse", "Cantidad"] = 4
# Recalcular porque cambió una entrada del cálculo.
productos["Ingreso"] = productos["Precio"] * productos["Cantidad"]
ordenados = productos.sort_values("Ingreso", ascending=False)
print(ordenados[["Producto", "Ingreso"]])

   Producto  Ingreso
p3  Monitor   199.99
p1    Mouse    63.96
p2  Teclado    51.00


**Qué observar:** Monitor 199.99; Mouse 63.96; Teclado 51.00.

**Recuerda:** Una columna calculada no es una fórmula de hoja de cálculo: si cambias sus entradas, debes volver a calcularla.

## Práctica
Intenta resolver antes de consultar las soluciones del aula web.

### Selecciona el monitor

Obtén el monitor con loc y luego con iloc. Consulta solo su precio.

**Criterio de logro:** 199.99 en ambos casos.

In [ ]:
# Escribe tu solución aquí.

### Compra de varias unidades

Filtra cantidades >= 2 y muestra productos ordenados por Precio de menor a mayor.

**Criterio de logro:** Mouse y Teclado, en ese orden.

In [ ]:
# Escribe tu solución aquí.

## Tarea · Consultas de librería

Construye una tabla: Cuaderno 1500×4, Lapicero 500×10, Carpeta 1200×3, Marcador 800×5, Regla 600×2.

1. Selecciona productos cuyo precio supere ₡700.
2. Crea Ingreso y suma el total.
3. Ordena por ingreso descendente en otra variable.
4. Escribe dos conclusiones. Control: total ₡19 800; mayor ingreso Cuaderno, ₡6000.

Evaluación: corrección 50 %, explicación 30 %, orden y reproducibilidad 20 %.

In [ ]:
# Desarrolla tu tarea en esta celda y agrega las que necesites.

## Mi conclusión
Escribe tu interpretación y una limitación.